## DATA WRANGLING AND CLEANING

*   Identify duplicate values in the dataset.

*   Remove duplicate values from the dataset.

*   Identify missing values in the dataset.

*   Impute the missing values in the dataset.

*   Normalize data in the dataset.


<hr>

In [1]:
import pandas as pd
import numpy as np
import os

Create file path and read files also create a DataFrame.

In [2]:
path = "C:\\Users\\Akshay\\Documents\\Data Analyst\\Developer project 1\\survey_results_public.csv"
df = pd.read_csv(path)
df.head()

ResponseId                                         MainBranch  \
0           1                     I am a developer by profession   
1           2             I am a student who is learning to code   
2           3  I am not primarily a developer, but I write co...   
3           4                     I am a developer by profession   
4           5                     I am a developer by profession   

                                          Employment  \
0  Independent contractor, freelancer, or self-em...   
1                                 Student, full-time   
2                                 Student, full-time   
3                                 Employed full-time   
4  Independent contractor, freelancer, or self-em...   

                                             Country US_State UK_Country  \
0                                           Slovakia      NaN        NaN   
1                                        Netherlands      NaN        NaN   
2                                 Russian Federation      NaN        NaN   
3                                            Austria      NaN        NaN   
4  United Kingdom of Great Britain and Northern I...      NaN    England   

                                             EdLevel     Age1stCode  \
0  Secondary school (e.g. American high school, G...  18 - 24 years   
1       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
2       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
3    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)  11 - 17 years   
4    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   5 - 10 years   

                                           LearnCode YearsCode  ...  \
0  Coding Bootcamp;Other online resources (ex: vi...       NaN  ...   
1  Other online resources (ex: videos, blogs, etc...         7  ...   
2  Other online resources (ex: videos, blogs, etc...       NaN  ...   
3                                                NaN       NaN  ...   
4                            Friend or family member        17  ...   

               Age Gender Trans                Sexuality  \
0  25-34 years old    Man    No  Straight / Heterosexual   
1  18-24 years old    Man    No  Straight / Heterosexual   
2  18-24 years old    Man    No        Prefer not to say   
3  35-44 years old    Man    No  Straight / Heterosexual   
4  25-34 years old    Man    No                      NaN   

                      Ethnicity                Accessibility  \
0  White or of European descent            None of the above   
1  White or of European descent            None of the above   
2             Prefer not to say            None of the above   
3  White or of European descent  I am deaf / hard of hearing   
4  White or of European descent            None of the above   

        MentalHealth           SurveyLength                  SurveyEase  \
0  None of the above  Appropriate in length                        Easy   
1  None of the above  Appropriate in length                        Easy   
2  None of the above  Appropriate in length                        Easy   
3                NaN  Appropriate in length  Neither easy nor difficult   
4                NaN  Appropriate in length                        Easy   

  ConvertedCompYearly  
0             62268.0  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  

[5 rows x 48 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83439 entries, 0 to 83438
Data columns (total 48 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   ResponseId                    83439 non-null  int64  
 1   MainBranch                    83439 non-null  object 
 2   Employment                    83323 non-null  object 
 3   Country                       83439 non-null  object 
 4   US_State                      14920 non-null  object 
 5   UK_Country                    4418 non-null   object 
 6   EdLevel                       83126 non-null  object 
 7   Age1stCode                    83243 non-null  object 
 8   LearnCode                     82963 non-null  object 
 9   YearsCode                     81641 non-null  object 
 10  YearsCodePro                  61216 non-null  object 
 11  DevType                       66484 non-null  object 
 12  OrgSize                       60726 non-null  object 
 13  C

Let's find *duplicate* values in a dataset

In [4]:
duplicate = df[df.duplicated()]
duplicate

Empty DataFrame
Columns: [ResponseId, MainBranch, Employment, Country, US_State, UK_Country, EdLevel, Age1stCode, LearnCode, YearsCode, YearsCodePro, DevType, OrgSize, Currency, CompTotal, CompFreq, LanguageHaveWorkedWith, LanguageWantToWorkWith, DatabaseHaveWorkedWith, DatabaseWantToWorkWith, PlatformHaveWorkedWith, PlatformWantToWorkWith, WebframeHaveWorkedWith, WebframeWantToWorkWith, MiscTechHaveWorkedWith, MiscTechWantToWorkWith, ToolsTechHaveWorkedWith, ToolsTechWantToWorkWith, NEWCollabToolsHaveWorkedWith, NEWCollabToolsWantToWorkWith, OpSys, NEWStuck, NEWSOSites, SOVisitFreq, SOAccount, SOPartFreq, SOComm, NEWOtherComms, Age, Gender, Trans, Sexuality, Ethnicity, Accessibility, MentalHealth, SurveyLength, SurveyEase, ConvertedCompYearly]
Index: []

[0 rows x 48 columns]

While the survey responses contain a wealth of information, we'll limit our analysis.
Let's select a subset of columns with the relevant data for our analysis.
- Demographic Information
- Education, Work, and Career
- Technology and Tech Culture

In [5]:
selected_columns = ["Country",
                    "Age",
                    "Gender",
                    "Accessibility",
                    "MentalHealth",
                    "MainBranch",
                    "Employment",
                    "EdLevel",
                    "Age1stCode",
                    "DevType",
                    "CompTotal",
                    "CompFreq",
                    "LanguageHaveWorkedWith",
                    "LanguageWantToWorkWith",
                    "DatabaseHaveWorkedWith",
                    "DatabaseWantToWorkWith",
                    "PlatformHaveWorkedWith",
                    "PlatformWantToWorkWith",
                    "WebframeHaveWorkedWith",
                    "WebframeWantToWorkWith",
                    "MiscTechHaveWorkedWith",
                    "MiscTechWantToWorkWith",
                    "ToolsTechHaveWorkedWith",
                    "ToolsTechWantToWorkWith",
                    "NEWCollabToolsHaveWorkedWith",
                    "NEWCollabToolsWantToWorkWith",
                    "OpSys",
                    "NEWStuck"
                   ]

In [6]:
len(selected_columns)

28

In [7]:
df = df[selected_columns].copy()

In [8]:
df.shape

(83439, 28)

Find *Null* values in dataset

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83439 entries, 0 to 83438
Data columns (total 28 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Country                       83439 non-null  object 
 1   Age                           82407 non-null  object 
 2   Gender                        82286 non-null  object 
 3   Accessibility                 77603 non-null  object 
 4   MentalHealth                  76920 non-null  object 
 5   MainBranch                    83439 non-null  object 
 6   Employment                    83323 non-null  object 
 7   EdLevel                       83126 non-null  object 
 8   Age1stCode                    83243 non-null  object 
 9   DevType                       66484 non-null  object 
 10  CompTotal                     47183 non-null  float64
 11  CompFreq                      52150 non-null  object 
 12  LanguageHaveWorkedWith        82357 non-null  object 
 13  L

Most columns have the data type object, either because they contain values of different types or contain empty values (NaN). It appears that every column contains some empty values since the Non-Null count for every column is lower than the total number of rows (83439). We'll need to deal with empty values and manually adjust the data type for each column on a case-by-case basis.

Find unique values in Age column

In [10]:
df["Age"].unique()

array(['25-34 years old', '18-24 years old', '35-44 years old',
       'Prefer not to say', '45-54 years old', 'Under 18 years old',
       '65 years or older', '55-64 years old', nan], dtype=object)

Replace *NaN* with Majority Age group.

In [11]:
majority = df["Age"].value_counts().idxmax()
df["Age"].replace(np.nan,majority,inplace=True)

In [12]:
df["Gender"].value_counts()

Man                                                                                   74817
Woman                                                                                  4120
Prefer not to say                                                                      1442
Non-binary, genderqueer, or gender non-conforming                                       690
Or, in your own words:                                                                  413
Man;Or, in your own words:                                                              268
Man;Non-binary, genderqueer, or gender non-conforming                                   252
Woman;Non-binary, genderqueer, or gender non-conforming                                 147
Man;Woman                                                                                41
Non-binary, genderqueer, or gender non-conforming;Or, in your own words:                 21
Man;Woman;Non-binary, genderqueer, or gender non-conforming                     

The gender column also allows for picking multiple options. We'll remove values containing more than one option to simplify our analysis.

In [13]:
df.where(~(df.Gender.str.contains(';|Or, in your own words:', na=False)), np.nan, inplace=True)

In [14]:
df["Gender"].value_counts()

Man                                                  74817
Woman                                                 4120
Prefer not to say                                     1442
Non-binary, genderqueer, or gender non-conforming      690
Name: Gender, dtype: int64

In [15]:
maj_Gender = df["Gender"].value_counts().idxmax()
df["Gender"].replace(np.nan,maj_Gender,inplace=True)

In [16]:
df["Employment"].unique()

array(['Independent contractor, freelancer, or self-employed',
       'Student, full-time', 'Employed full-time', 'Student, part-time',
       'I prefer not to say', 'Employed part-time',
       'Not employed, but looking for work', 'Retired', nan,
       'Not employed, and not looking for work'], dtype=object)

Replace NaN values with Max frequency.

In [17]:
maj_grp = df["Employment"].value_counts().idxmax()
df["Employment"].replace(np.nan,maj_grp,inplace=True)

In [18]:
print(df["EdLevel"].unique())
print(df["Age1stCode"].unique())

['Secondary school (e.g. American high school, German Realschule or Gymnasium, etc.)'
 'Bachelor’s degree (B.A., B.S., B.Eng., etc.)'
 'Master’s degree (M.A., M.S., M.Eng., MBA, etc.)'
 'Other doctoral degree (Ph.D., Ed.D., etc.)'
 'Some college/university study without earning a degree' 'Something else'
 'Professional degree (JD, MD, etc.)' 'Primary/elementary school'
 'Associate degree (A.A., A.S., etc.)' nan]
['18 - 24 years' '11 - 17 years' '5 - 10 years' '25 - 34 years' nan
 '35 - 44 years' 'Younger than 5 years' '45 - 54 years' '55 - 64 years'
 'Older than 64 years']


Replace NaN values with value of a maximum count. Find maximum count using *idxmax()*.

In [19]:
max_count_EdLevel = df["EdLevel"].value_counts().idxmax()
print(max_count_EdLevel)
max_count_Age1stCode = df["Age1stCode"].value_counts().idxmax()
print(max_count_Age1stCode)

Bachelor’s degree (B.A., B.S., B.Eng., etc.)
11 - 17 years


In [20]:
df["EdLevel"].replace(np.nan, max_count_EdLevel, inplace=True)
df["Age1stCode"].replace(np.nan, max_count_Age1stCode, inplace=True)

###### DATA NORMALIZATION

In [21]:
df[["CompFreq","CompTotal"]]

CompFreq  CompTotal
0      Monthly     4800.0
1          NaN        NaN
2          NaN        NaN
3      Monthly        NaN
4          NaN        NaN
...        ...        ...
83434   Yearly   160500.0
83435  Monthly   200000.0
83436   Weekly     1800.0
83437  Monthly    90000.0
83438  Monthly     7700.0

[83439 rows x 2 columns]

List out the various categories in the column 'CompFreq'.

In [22]:
max_count = df["Country"].value_counts().idxmax()
df["Country"].replace(np.nan, max_count, inplace=True)

In [23]:
df["CompFreq"].value_counts()

Yearly     25454
Monthly    24491
Weekly      1595
Name: CompFreq, dtype: int64

There are two columns in the dataset that talk about compensation.

One is "CompFreq". This column shows how often a developer is paid (Yearly, Monthly, Weekly).

The other is "CompTotal". This column talks about how much the developer is paid per Year, Month, or Week depending upon his/her "CompFreq".

This makes it difficult to compare the total compensation of the developers.

In this section you will create a new column called 'NormalizedAnnualCompensation' which contains the 'Annual Compensation' irrespective of the 'CompFreq'.

Once this column is ready, it makes comparison of salaries easy.

In [24]:
df["NormalizedAnnualCompensation"] = np.where(
    df["CompFreq"] == "Weekly",df["CompTotal"]*52,np.where(
    df["CompFreq"] == "Monthly",df["CompTotal"]*12,df["CompTotal"]))
df.head()

Country              Age Gender  \
0                                           Slovakia  25-34 years old    Man   
1                                        Netherlands  18-24 years old    Man   
2                                 Russian Federation  18-24 years old    Man   
3                                            Austria  35-44 years old    Man   
4  United Kingdom of Great Britain and Northern I...  25-34 years old    Man   

                 Accessibility       MentalHealth  \
0            None of the above  None of the above   
1            None of the above  None of the above   
2            None of the above  None of the above   
3  I am deaf / hard of hearing                NaN   
4            None of the above                NaN   

                                          MainBranch  \
0                     I am a developer by profession   
1             I am a student who is learning to code   
2  I am not primarily a developer, but I write co...   
3                     I am a developer by profession   
4                     I am a developer by profession   

                                          Employment  \
0  Independent contractor, freelancer, or self-em...   
1                                 Student, full-time   
2                                 Student, full-time   
3                                 Employed full-time   
4  Independent contractor, freelancer, or self-em...   

                                             EdLevel     Age1stCode  \
0  Secondary school (e.g. American high school, G...  18 - 24 years   
1       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
2       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
3    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)  11 - 17 years   
4    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   5 - 10 years   

                                             DevType  ...  \
0                                  Developer, mobile  ...   
1                                                NaN  ...   
2                                                NaN  ...   
3                               Developer, front-end  ...   
4  Developer, desktop or enterprise applications;...  ...   

   WebframeWantToWorkWith                 MiscTechHaveWorkedWith  \
0                     NaN                                    NaN   
1                     NaN                                Cordova   
2                   Flask  NumPy;Pandas;TensorFlow;Torch/PyTorch   
3          Angular;jQuery                                    NaN   
4                   Flask       Apache Spark;Hadoop;NumPy;Pandas   

                        MiscTechWantToWorkWith     ToolsTechHaveWorkedWith  \
0                                          NaN                         NaN   
1                                          NaN             Docker;Git;Yarn   
2  Keras;NumPy;Pandas;TensorFlow;Torch/PyTorch                         NaN   
3                                          NaN                         NaN   
4                          Hadoop;NumPy;Pandas  Docker;Git;Kubernetes;Yarn   

      ToolsTechWantToWorkWith  \
0                         NaN   
1                         Git   
2                         NaN   
3                         NaN   
4  Docker;Git;Kubernetes;Yarn   

                        NEWCollabToolsHaveWorkedWith  \
0                                     PHPStorm;Xcode   
1          Android Studio;IntelliJ;Notepad++;PyCharm   
2  IPython/Jupyter;PyCharm;RStudio;Sublime Text;V...   
3                                                NaN   
4         Atom;IPython/Jupyter;Notepad++;PyCharm;Vim   

                        NEWCollabToolsWantToWorkWith        OpSys  \
0                                         Atom;Xcode        MacOS   
1                                                NaN      Windows   
2  IPython/Jupyter;RStudio;Sublime Text;Visual St...        MacOS   
3                                                NaN      Windows   
4  Atom;IPython/Ju

We've now cleaned up and prepared the dataset for analysis. Let's take a look at a sample of rows from the data frame.

In [25]:
df.sample(10)

Country                 Age Gender  \
74801                      Ethiopia  Under 18 years old    Man   
29573                   Netherlands     25-34 years old    Man   
32942                         India     35-44 years old    Man   
62042                     Guatemala     35-44 years old    Man   
30443                     Australia     45-54 years old    Man   
15424  Iran, Islamic Republic of...     18-24 years old    Man   
29940      United States of America     35-44 years old    Man   
68416                       Germany     25-34 years old    Man   
77554                         Japan     25-34 years old    Man   
33156                         India     18-24 years old    Man   

           Accessibility                                       MentalHealth  \
74801  None of the above                                                NaN   
29573  None of the above                                  None of the above   
32942  None of the above  I have a concentration and/or memory disorder ...   
62042  None of the above                                  None of the above   
30443                NaN                                                NaN   
15424  None of the above                                  None of the above   
29940  None of the above                                  None of the above   
68416  None of the above                                  None of the above   
77554  None of the above  I have a concentration and/or memory disorder ...   
33156  None of the above                                  None of the above   

                                              MainBranch  \
74801             I am a student who is learning to code   
29573                     I am a developer by profession   
32942                     I am a developer by profession   
62042                     I am a developer by profession   
30443                     I am a developer by profession   
15424                     I am a developer by profession   
29940  I am not primarily a developer, but I write co...   
68416                     I am a developer by profession   
77554                     I am a developer by profession   
33156             I am a student who is learning to code   

                                              Employment  \
74801                                 Student, part-time   
29573                                 Employed full-time   
32942                                 Employed full-time   
62042                                 Employed full-time   
30443                                 Employed full-time   
15424                                 Student, part-time   
29940                                 Employed full-time   
68416                                 Employed full-time   
77554  Independent contractor, freelancer, or self-em...   
33156                                 Student, full-time   

                                                 EdLevel     Age1stCode  \
74801                          Primary/elementary school  11 - 17 years   
29573       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
32942       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  18 - 24 years   
62042       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   5 - 10 years   
30443         Other doctoral degree (Ph.D., Ed.D., etc.)  11 - 17 years   
15424  Secondary school (e.g. American high school, G...   5 - 10 years   
29940    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   5 - 10 years   
68416       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  18 - 24 years   
77554    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)  11 - 17 years   
33156       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   5 - 10 years   

                                                 DevType  ...  \
74801                                                NaN  ...   
29573  Developer, mobile;Developer, front-end;Develop...  ...   
32942                                Developer, back-end  ...   
62042  Developer, full-

In [26]:
df.to_csv("C:\\Users\\Akshay\\Documents\\Data Analyst\\Developer project 1\\Cleaned_Data.csv")

### Thank You :) 